core/contacts.go

```diff
@@ -136,95 +164,119 @@ func (t *Textile) UpdateContactInboxes() error {
 	return t.datastore.Contacts().UpdateInboxes(t.node.Identity.Pretty(), inboxes)
 }
 
-// FindContact searches the network for contacts via cafe sessions
-func (t *Textile) FindContact(query *ContactInfoQuery) (*ContactInfoQueryResult, error) {
-	if query.Add && query.Username != "" {
-		return nil, fmt.Errorf("add not allowed when searching by username")
-	}
+// FindContacts searches the network for contacts
+func (t *Textile) FindContacts(query *ContactQuery) (<-chan *ContactQueryResult, <-chan error, *broadcast.Broadcaster) {
+	set := newContactSet()
+	var searchChs []chan *ContactQueryResult
 
-	result := &ContactInfoQueryResult{
-		Local:  make([]ContactInfo, 0),
-		Remote: make([]ContactInfo, 0),
-	}
+	// local results channel
+	localCh := make(chan *ContactQueryResult)
+	searchChs = append(searchChs, localCh)
 
-	// search local
-	for _, c := range t.datastore.Contacts().Find(query.Id, query.Address, query.Username) {
-		i := t.contactInfo(&c, true)
-		if i != nil {
-			result.Local = append(result.Local, *i)
+	// remote results channel(s)
+	var cafeChs []chan *ContactQueryResult
+	clientCh := make(chan *ContactQueryResult)
+	sessions := t.datastore.CafeSessions().List()
+	if len(sessions) > 0 {
+		for range sessions {
+			cafeCh := make(chan *ContactQueryResult)
+			cafeChs = append(cafeChs, cafeCh)
+			searchChs = append(searchChs, cafeCh)
 		}
+	} else {
+		searchChs = append(searchChs, clientCh)
 	}
 
-	if query.Local || len(result.Local) >= query.Limit {
-		return result, nil
-	}
+	resultCh := mergeContactQueryResults(searchChs)
```

Comment:
`resultCh` is a merge of all the concurrent search results...

Reply:
So cool